### PR TITLE
Engine: Iterating Hash Table

### PR DESCRIPTION
**Pull request by sherry-1001**
### What problem does this PR solve?
Add features about iterating hash table.
Scanning hash table using iterator, similarly like STL iterator.  And support range  iterator.
If need to lock slot during iterating hash table, developer can use range iterator.  Please see example in test/tests.cpp: `TestHashTableIterator` case.

Issue Number: [197](https://github.com/pmem/kvdk/issues/197)

Problem Summary:
   Iterating Hash Table is a basic feature for backup and cleaning expired kv, and other functions.

Tests

- Unit test
    `TestHashTableIterator` case
    `TestHashTableRangeIter` case
- Integration test
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
    - Consumes more PMEM

### CODE REVIEW

File: engine/kv_engine.hpp [hunk 1]
```diff
@@ -114,6 +114,8 @@ class KVEngine : public Engine {
     return skiplists_;
   };
 
+  const std::shared_ptr<HashTable> GetHashTable() { return hash_table_; }
```
Comment (ZiyanShi) [inline engine/kv_engine.hpp:117]: Don't expose internal components. Use special macros to expose them for test builds.
Also, no need to return const.

File: engine/hash_table.hpp [hunk 2]
```diff
@@ -164,6 +166,101 @@ class HashTable {
...
   }
 
+  class HashTableIterator {
+   private:
+    friend class HashTable;
+
+    HashTable* hash_table_;
+    uint64_t bucket_idx_;
+    uint64_t entry_idx_;
+    char* bucket_ptr_;
```
Comment (ZiyanShi) [inline engine/hash_table.hpp:176]: HashEntry*

File: engine/hash_table.hpp [hunk 3]
```diff
@@ -164,6 +166,101 @@ class HashTable {
...
+             !hash_table_->hash_bucket_entries_[bucket_idx_]) {
+        bucket_idx_++;
+      }
+      if (bucket_idx_ == hash_table_->iter_end_bucket_idx_) {
+        bucket_ptr_ = nullptr;
+        return;
+      }
+      bucket_ptr_ = (char*)hash_table_->main_buckets_ +
+                    bucket_idx_ * hash_table_->hash_bucket_size_;
+      _mm_prefetch(bucket_ptr_, _MM_HINT_T0);
```
Comment (ZiyanShi) [inline engine/hash_table.hpp:229]: Software prefetches usually does not help.

**Review by ZiyanShi — APPROVED** [review]

File: engine/hash_table.hpp (hunk 2 above)
Comment (sherry-1001) [reply engine/hash_table.hpp:232]: can't use hash entry

File: engine/hash_table.hpp [hunk 4]
```diff
@@ -74,6 +74,8 @@ struct alignas(16) HashEntry {
 
   bool Empty() { return header_.index_type == HashIndexType::Empty; }
 
+  bool Invalid() { return header_.index_type == HashIndexType::Invalid; }
```
Comment (ZiyanShi) [inline engine/hash_table.hpp:77]: Is this used?

Comment (ZiyanShi) [inline engine/hash_table.hpp:77]: Is this used?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

**Review by ZiyanShi — APPROVED** [review]

**Review by JiayuZzz — APPROVED** [review]
LGTM

File: engine/hash_table.hpp [hunk 5]
```diff
@@ -197,4 +202,126 @@ class HashTable {
...
+  uint64_t iter_start_bucket_idx_;
+  uint64_t iter_end_bucket_idx_;
+  // lock current access slot
+  std::unique_lock<SpinMutex> iter_lock_slot_;
+  // current slot id
+  uint64_t current_slot_id;
+  HashTable* hash_table_;
+
+  void GetBucketRangeAndLockSlot() {
+    std::unique_lock<SpinMutex> lock_slot{
```
Comment (ZiyanShi) [inline engine/hash_table.hpp:220]: may dead lock with other operations.
Release iter_lock_slot_ before lock another.

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**